synapse/storage/prepare_database.py

```diff
@@ -413,6 +422,24 @@ def _upgrade_existing_database(
         start_ver += 1
 
     logger.debug("applied_delta_files: %s", current_schema_state.applied_deltas)
+    logger.debug(
+        "pending background_updates: %s",
+        (name for name, ordering in current_schema_state.background_updates),
+    )
+
+    # Bail if there are any pending background updates from before the background schema compat version.
+    for update_name, ordering in sorted(
+        current_schema_state.background_updates, key=lambda b: b[1]
+    ):
+        # ordering is an int based on when the background update was added:
+        #
+        #   (schema version when added * 100) + (schema delta when added).
+        update_schema_version = ordering // 100
```

Comment:
I'm a bit concerned that the fact that the ordering is `(schema version when added * 100) + N` is a bit of a loose convention.

In many ways I'd find it *less* confusing if, instead of using a `COMPAT_VERSION` following the same system as `SCHEMA_VERSION` etc, we instead had a `MIN_PENDING_BACKGROUND_UPDATE_ORDER` which we compared with the raw `ordering` of the pending updates.

Reply:
It is definitely a bit loose, using the raw minimum might make more sense.